libexecutive/Executive.cpp

```diff
@@ -382,6 +367,62 @@ bool Executive::executeCreate(Address const& _sender, u256 const& _endowment, u2
     return !m_ext;
 }
 
+void Executive::grantContractStatusManager(TableFactory::Ptr memoryTableFactory,
+    Address const& newAddress, Address const& sender, Address const& origin)
+{
+    LOG(TRACE) << LOG_DESC("grantContractStatusManager") << LOG_KV("contract", newAddress)
+               << LOG_KV("sender account", sender) << LOG_KV("origin account", origin);
+
+    std::string tableName = precompiled::getContractTableName(newAddress);
+    auto table = memoryTableFactory->openTable(tableName);
+
+    if (!table)
+    {
+        LOG(ERROR) << LOG_DESC("grantContractStatusManager get newAddress table error!");
+        return;
+    }
+
+    // grant origin authorization
+    auto entry = table->newEntry();
+    entry->setField("key", "authority");
+    entry->setField("value", origin.hex());
+    table->insert("authority", entry);
+
+    if (origin != sender)
+    {
+        // the authority field is updated to add the sender's authorization
+        std::string senderTableName = precompiled::getContractTableName(sender);
+        auto senderTable = memoryTableFactory->openTable(senderTableName);
+        if (!senderTable)
+        {
+            LOG(ERROR) << LOG_DESC("grantContractStatusManager get sender table error!");
+            return;
+        }
+
+        auto entries = senderTable->select("authority", senderTable->newCondition());
+        if (entries->size() == 0)
+        {
+            LOG(ERROR) << LOG_DESC("grantContractStatusManager no sender authority is granted");
+            return;
+        }
+        std::string authority = entries->get(0)->getField("value");
+        if (authority.find(origin.hex()) == std::string::npos)
+        {
+            // remove duplicate
+            authority += "," + origin.hex();
+        }
+        auto entry = table->newEntry();
+        entry->setField("key", "authority");
+        entry->setField("value", authority);
+        table->update("authority", entry, table->newCondition());
+        LOG(TRACE) << LOG_DESC("grantContractStatusManager") << LOG_KV("authorityNew", authority);
```

Comment:
Recommend to use `DEBUG` log since it's a low-frequency operation.

Reply:
Done